crates/amf0/src/ser.rs

```diff
@@ -0,0 +1,948 @@
+//! Serialize a Rust data structure into AMF0 data.
+
+use std::io;
+
+use serde::Serialize;
+use serde::ser::{
+    Impossible, SerializeMap, SerializeSeq, SerializeStruct, SerializeStructVariant, SerializeTuple, SerializeTupleStruct,
+    SerializeTupleVariant,
+};
+
+use crate::Amf0Error;
+use crate::encoder::Amf0Encoder;
+
+/// Serialize a value into a given writer.
+pub fn to_writer<W>(writer: W, value: &impl serde::Serialize) -> crate::Result<()>
+where
+    W: io::Write,
+{
+    let mut serializer = Serializer::new(writer);
+    value.serialize(&mut serializer)
+}
+
+/// Serialize a value into a new byte vector.
+pub fn to_bytes(value: &impl serde::Serialize) -> crate::Result<Vec<u8>> {
+    let mut writer = Vec::new();
+    to_writer(&mut writer, value)?;
+    Ok(writer)
+}
+
+/// Serializer for AMF0 data.
+pub struct Serializer<W> {
+    encoder: Amf0Encoder<W>,
+}
+
+impl<W> Serializer<W> {
+    /// Create a new AMF0 serializer from a writer.
+    pub fn new(writer: W) -> Self {
+        Self {
+            encoder: Amf0Encoder::new(writer),
+        }
+    }
+}
```

Comment:
Does this need to be a new type? can we not just impl serializer for `&Mut amf0Encoder`?